File: beakjoon/2156.py
import sys

sys.setrecursionlimit(15000)
readline = sys.stdin.readline


def findMaxWineAmt(idx):
    if idx >= n:
        return 0
    if cache[idx] != -1:
        return cache[idx]

    ret = max(findMaxWineAmt(idx + 1), wineAmt[idx] + findMaxWineAmt(idx + +2))
    if idx <= n - 2:
        ret = max(ret, wineAmt[idx] + wineAmt[idx + 1] + findMaxWineAmt(idx + 3))
    cache[idx] = ret
    return ret


wineAmt = []
n = int(readline())
for _ in range(n):
    wineAmt.append(int(readline()))
cache = [-1] * n
print(findMaxWineAmt(0))

File: beakjoon/2156.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

int n;
int arr[10003];
int cache[10003][3];

int dp(int k,int p) {
	if (k >= n) return 0;
	int &ret = cache[k][p];
	if (ret != 0) return ret;

	if (p == 2) return ret = dp(k + 1, 0);
	return ret = max(dp(k + 1, p + 1) + arr[k], dp(k + 1, 0));
}

int main()
{
	scanf("%d", &n);
	for (int i = 0; i < n; ++i) scanf("%d", arr + i);
	printf("%d\n", dp(0, 0));
	
}